[WIP] Rackspace Cloud support

thommay opened the pull request:
This PR hooks up gophercloud for Rackspace support. It successfully finds minions and creates/updates/deletes Load Balancers.
What it does not do is route traffic from the LB to the instance; Rackspace also uses a NATing LB. (it probably doesn't have working tests right now, either)

thommay committed:
initial rackspace support

thommay committed:
working LB creation

thommay commented:
In terms of LB support, the situation is slightly different to AWS. Here, at least, we know the source IP that traffic will come from the LB to the node, so with some modification to the proxy we should be able to  cause traffic from that source to our port to be correctly directed.

jlowdermilk commented:
cc @brendandburns. @thockin, please reassign if there's someone better equipped to review.

thockin commented:
Regarding load-balancing working differently that GCE and AWS - I knew that
would pop up at some point.

Do you have some packet dumps or something that will give some confidence
in a plan?  Does forwarded traffic arrive with source-ip set to the load
balancer's IP?  I'm assuming it's too much to expect the port to be set...

We could do something like capture traffic from the public IP bound for the
service port and assume that load-balancers don't balance for multiple
services at once.  I am not sure what will happen if we set both captures
rules - to and from the public IP - in other environments.  I'm
specifically thinking that it will cause a loop in GCE and that it will
totally destroy simpler environments like digital ocean.

We might have to finally give in and make this a function of the
cloud-provider API.  Ugh.

Can these balancers target pod IPs directly or can they only hit VM IPs
(GCE has this limitation)?

I'd be happy to discuss this issue further.

As for the rest of the PR it looks fine to me.  I mean, the kubernetes
parts are all "yours" and the dep is too.

On Fri, Feb 6, 2015 at 10:28 AM, Jeff Lowdermilk <notifications@github.com>
wrote:

> Assigned #4215
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4215> to @thockin
> <https://github.com/thockin>.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4215#event-231658337>
> .
>

thommay commented:
In this case we can definitely assume that the LB doesn't deal with more than one service at once - rackspace LBs can only be bound to a single external port. 
In terms of a concrete example, when we create an LB we get the following from the API:
```
LB Source Address is {IPv4Public:104.130.254.5 IPv4Private:10.189.252.5 IPv6Public:2001:4800:7904:100::2/64 IPv6Private:}
```
If we then do a packet capture on the private network, we see:
```
11:00:00.859577 IP 10.189.252.5.47885 > 10.209.0.214.tram: Flags [.], ack 1780561873, win 29, options [nop,nop,TS val 254063292 ecr 373224474], length 0
11:00:02.204288 IP 10.189.252.5.47885 > 10.209.0.214.tram: Flags [F.], seq 0, ack 1, win 29, options [nop,nop,TS val 254063628 ecr 373224474], length 0
11:00:02.206915 IP 10.189.252.5.47885 > 10.209.0.214.tram: Flags [R], seq 3018979592, win 0, length 0
11:00:02.206938 IP 10.189.252.5.47885 > 10.209.0.214.tram: Flags [R], seq 3018979592, win 0, length 0
11:00:02.331993 IP 10.189.252.5.47937 > 10.209.0.214.tram: Flags [S], seq 3483766675, win 14600, options [mss 1460,sackOK,TS val 254063660 ecr 0,nop,wscale 9], length 0
11:00:02.334304 IP 10.189.252.5.47937 > 10.209.0.214.tram: Flags [.], ack 834631768, win 29, options [nop,nop,TS val 254063660 ecr 373225948], length 0
```
and if we then add:
```
sudo iptables -t nat -A KUBE-PORTALS-CONTAINER -s 10.189.252.5/32 -p tcp -m comment --comment rubyservice -m tcp --dport 4567 -j DNAT --to-destination 10.209.0.214:58466
sudo iptables -t nat -A KUBE-PORTALS-HOST -s 10.189.252.5/32 -p tcp -m comment --comment rubyservice -m tcp --dport 4567 -j DNAT --to-destination 10.209.0.214:58466
```
to our minion, we get a working Load Balancer.

thommay commented:
So my thought was that we know up front that, eg, Rackspace is a NATing LB, but GCE is a pass-through. So rather than just returning an IP from CreateTCPLoadBalancer, we return a struct that is something like
```go
{
  DestIP string
  SourceIP string
  Rewrite bool
}
```
and if we then pass that through to the proxy, it can write either source captures or destination captures based on the Rewrite field.

thommay committed:
Support packet rewriting load balancers

This introduces a flag to allow cloud providers to signal that their
load balancers are rewriting (NATing) incoming packets. This allows us
to support Rackspace cloud LBs out of the box, and should make it
easier to handle AWS down the line.

thommay commented:
@thockin PTAL

thockin commented on pkg/api/types.go:
```diff
@@ -683,6 +683,14 @@ const (
+type LoadBalancerInfo struct {
+	DestIP   string `json:"destIP,omitempty" description:"External IP of the Load Balancer"`
+	SourceIP string `json:"sourceIP,omitempty" description:"Source IP of the Load Balancer"`
+	Rewrite  bool   `json:"rewrite,omitempty" description:"Does the LB rewrite packets"`
```
If we're going to do this - and I still need to think about it some more - we should define this as a string type and define "Forwarding" and "BackToBack" for GCE and Rackspace respectively.

@brendanburns we talked about AWS, but I forgot what you said - thoughts on this?

thockin commented on pkg/api/types.go:
```diff
@@ -683,6 +683,14 @@ const (
+// pass through
+type LoadBalancerInfo struct {
+	DestIP   string `json:"destIP,omitempty" description:"External IP of the Load Balancer"`
+	SourceIP string `json:"sourceIP,omitempty" description:"Source IP of the Load Balancer"`
```
Why two fields for IP?

thommay replied:
Well, mostly for reporting. It's pretty miserable to have to log in to the rackspace management console (or whichever) - and users may not have access to that - to get the address of the loadbalancer.

thockin replied:
Ahh, I see.  The IP of packets from the LB is not even the LB's public IP.  Double puke.  This makes me unhappy.  I don't have a better answer on hand, but I really want to think about this more.  What would you say to making two PRs out of this - one for all your support EXCEPT public LB, and a second for just that part?

thommay replied:
@thockin done in #4478 for the base line. I'll try rebasing this onto that one to preserve the history

thockin replied:
Thanks.

On Tue, Feb 17, 2015 at 3:03 AM, Thom May <notifications@github.com> wrote:

> In pkg/api/types.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4215#discussion_r24807709>
> :
>
> > @@ -683,6 +683,14 @@ const (
> >  	AffinityTypeNone AffinityType = "None"
> >  )
> >
> > +// LoadBalancerInfo defines whether the LoadBalancer is rewriting or
> > +// pass through
> > +type LoadBalancerInfo struct {
> > +	DestIP   string `json:"destIP,omitempty" description:"External IP of the Load Balancer"`
> > +	SourceIP string `json:"sourceIP,omitempty" description:"Source IP of the Load Balancer"`
>
> @thockin <https://github.com/thockin> done in #4478
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4478> for the
> base line. I'll try rebasing this onto that one to preserve the history
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4215/files#r24807709>
> .
>

thommay committed:
Basic Rackspace cloud support

This enables all but Load Balancer support for the Rackspace public
cloud platform.

erictune commented:
Please separate into a Godeps PR and an everything-else PR.

erictune commented:
Please consider updating docs/getting-started-guides/README.md to add yourself as a maintainer for rackspace.

thommay commented:
I'm going to close this since #4478 landed the GoDeps changes, and #4480 has the discussion for LBs
